tuf/ngclient/_internal/trusted_metadata_set.py

```diff
@@ -0,0 +1,464 @@
+# Copyright the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Trusted collection of client-side TUF Metadata
+
+TrustedMetadataSet keeps track of current valid set of metadata for the client,
+and handles almost every step of the "Detailed client workflow" (
+https://theupdateframework.github.io/specification/latest#detailed-client-workflow)
+in the TUF specification: the remaining steps are related to filesystem and
+network IO which is not handled here.
+
+Loaded metadata can be accessed via the index access with rolename as key
+(trusted_set["root"]) or, in the case of top-level metadata using the helper
+properties (trusted_set.root).
+
+The rules for top-level metadata are
+ * Metadata is loadable only if metadata it depends on is loaded
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Metadata must be loaded/updated in order:
+   root -> timestamp -> snapshot -> targets -> (other delegated targets)
+
+
+Exceptions are raised if metadata fails to load in any way.
+
+Example of loading root, timestamp and snapshot:
+
+>>> # Load local root (RepositoryErrors here stop the update)
+>>> with open(root_path, "rb") as f:
+>>>     trusted_set = TrustedMetadataSet(f.read())
+>>>
+>>> # update root from remote until no more are available
+>>> with download("root", trusted_set.root.signed.version + 1) as f:
+>>>     trusted_set.update_root(f.read())
+>>> # ...
+>>> trusted_set.root_update_finished()
+>>>
+>>> # load local timestamp, then update from remote
+>>> try:
+>>>     with open(timestamp_path, "rb") as f:
+>>>         trusted_set.update_timestamp(f.read())
+>>> except (RepositoryError, OSError):
+>>>     pass # failure to load a local file is ok
+>>>
+>>> with download("timestamp") as f:
+>>>     trusted_set.update_timestamp(f.read())
+>>>
+>>> # load local snapshot, then update from remote if needed
+>>> try:
+>>>     with open(snapshot_path, "rb") as f:
+>>>         trusted_set.update_snapshot(f.read())
+>>> except (RepositoryError, OSError):
+>>>     # local snapshot is not valid, load from remote
+>>>     # (RepositoryErrors here stop the update)
+>>>     with download("snapshot", version) as f:
+>>>         trusted_set.update_snapshot(f.read())
```

Comment:
```suggestion
>>> # update snapshot from remote
>>> with download("snapshot", version) as f:
>>>     trusted_set.update_snapshot(f.read())
```
Same suggestion here, does it make sense to simplify the example by only fetching from the remote?